### ISSUE TITLE
Make use of run_in_tmpdir argument for tests

### ISSUE THREAD

**Issue by RemDelaporteMathurin** [festim-dev/FESTIM#530]
Currently in the tests we have a lot of custom paths made from the temp_directory. But with pytest, one can simply add run_in_tmpdir in the arguments of the test.

**Comment by KulaginVladimir** [comment]
@RemDelaporteMathurin, I'd like to refactor tests so they, at least, do not [produce ](https://github.com/festim-dev/FESTIM/issues/693) unneeded files. Maybe, some additional improvements are needed, e.g. a unified code style for tests, filling of missed docstrings or smth else?

**Comment by RemDelaporteMathurin** [comment]
The tests would greatly benefit from refactoring and tidying up.

Mostly I think the directory architecture needs rework and instead of having unit, integration, system etc, we have just files like "test_temperature.py" "test_simulation.py" etc.

I would like to keep all the MMS tests in one file though "test_mms.py".

In addition to that, I'm sure some tests miss docstrings yes.
Since we didn't know of `parametrize` when we started, some tests could potentially benefit from it too.

It's a lot to do though! 😄 

I agree the tests should at least not produce all these files.

**Comment by KulaginVladimir** [comment]
I suppose that reworking of the directory architecture is a tough point (at least for me). Possibly, a specific sketch of a new architecture might help.

Will it be OK for the first step if I start refactoring the tests. Also I think some specific type of tests (as you mentioned MMS) can be collected to the dedicated files strightforward.

**Comment by RemDelaporteMathurin** [comment]
Yes that's fine with me!

**Comment by RemDelaporteMathurin** [comment]
In this specific case this is fine I think, these two assert statements are effectively testing the same thing.

**Comment by KulaginVladimir** [comment]
For these [tests](https://github.com/festim-dev/FESTIM/blob/fdfcb17465e12ac74bff76668d646aeaa6f2d30a/test/unit/test_helpers.py#L29-L63), can we define `CustomExpr` separately and call it in tests, parametrised with pytest?

**Comment by KulaginVladimir** [comment]
Also, I think [these ](https://github.com/festim-dev/FESTIM/blob/fdfcb17465e12ac74bff76668d646aeaa6f2d30a/test/unit/test_sources.py#L51-L93) tests can be parametrised as well as individual test files for derived quantitites can be collected into several parametrised tests. However, I have some doubts on the general refactoring of tests considering the questions adressed on the meeting.

**Comment by RemDelaporteMathurin** [comment]
> For these [tests](https://github.com/festim-dev/FESTIM/blob/fdfcb17465e12ac74bff76668d646aeaa6f2d30a/test/unit/test_helpers.py#L29-L63), can we define `CustomExpr` separately and call it in tests, parametrised with pytest?

Yes


> Also, I think [these ](https://github.com/festim-dev/FESTIM/blob/fdfcb17465e12ac74bff76668d646aeaa6f2d30a/test/unit/test_sources.py#L51-L93) tests can be parametrised

Yes



> as well as individual test files for derived quantitites can be collected into several parametrised tests.

Not sure I understand this

**Comment by KulaginVladimir** [comment]
For example, `test_minimum_surface`:
https://github.com/festim-dev/FESTIM/blob/dfec3779e50a2dcd3266459e042f76985683e686/test/unit/test_exports/test_derived_quantities/test_minimum_surface.py#L6-L17
Now it has `test_title_*` tests that surely can be parametrised to smth like this:
```python
@pytest.mark.parametrize("field,surface", [("solute", 1), ("T", 2)])
def test_title(field, surface):
    """
    A simple test to check that the title is set
    correctly in festim.MinimumSurface

    Args:
        field (str, int):  the field ("solute", 0, 1, "T", "retention")
        surface (int): the surface id
    """

    my_min = MinimumSurface(field, surface)
    assert my_min.title == "Minimum {} surface {}".format(field, surface)
``` 
Such `test_title_*` test exists in all other files for tests of derived quantities. I thought to add another parametrisation to swap between `DerivedQuantity` objects, e.g. `MaximumSurface`, `MinimalSurface`, `AverageSurface`.

**Comment by RemDelaporteMathurin** [comment]
The titles are not super consistent from one derived quantity to another. I think @jhdark has a PR to fix it (#711) that he needs to reopen.

Yes, parametrise them by field and surface is a good idea! I don't know yet about parametrising for the quantity type since they are not yet consistent with each other (sometimes it's "Minimum {} surface {}" sometimes it's "Surface flux {}: {}" for instance)

**Comment by RemDelaporteMathurin** [comment]
I think we can close this!
